gadgets/src/less_than.rs

```diff
@@ -81,7 +86,21 @@ impl<F: Field, const N_BYTES: usize> LtChip<F, N_BYTES> {
                 .map(move |poly| q_enable.clone() * poly)
         });
 
-        LtConfig { lt, diff, range }
+        meta.annotate_lookup_any_column(u8, || "LOOKUP_u8");
+
+        diff[0..N_BYTES].iter().for_each(|column| {
+            meta.lookup_any("range check lookup", |meta| {
+                let exp = meta.query_advice(*column, Rotation::cur());
+                vec![(exp, meta.query_fixed(u8, Rotation::cur()))]
+            });
+        });
```

Comment:
I have two suggestions here:

1. What do you think we merge them into one gate? It looks more readable to me if one purpose is achieved using only one gate.
2. It looks like the term "exp" is copied from the exp circuit. We can rename it to u8-related terms.

```suggestion
        meta.lookup_any("range check for u8", |meta| {
            diff[0..N_BYTES]
                .iter()
                .map(|column| {
                    let u8_cell = meta.query_advice(*column, Rotation::cur());
                    let u8_range = meta.query_fixed(u8, Rotation::cur());
                    (u8_cell, u8_range)
                })
                .collect()
        });
```

Reply:
I followed your advice as the two expressions seem equivalent to me. The only difference is that expression 1 creates a vector with a single tuple for each element in the iterator, while Expression 2 creates a tuple for each element and then collects all tuples into a collection.

Despite that, the version containing the commit with the updated syntax generates an error in the range check (as you can see below). Can you explain why?

Reply:
I'm so sorry. I made a mistake. Your previous implementation was correct. My suggestion was wrong. 
My proposed constraints would enforce all cells to be the same value.

Think about looking up the expression `A xor B = C`. You would program the gate to return `vec![(cell_a, table_col_a), (cell_b, table_col_b), (cell_c, table_col_c)]`. That means cells a, b, and c must be on the same row of the table columns.

Now say we have cells `u8_a` and `u8_b`. Writing two gates constrain them to be within u8 range, but `a` and `b` could differ. Writing one gate that returns a vector of two elements adds an additional constraint that a and b have to be equal.

Reply:
No problem at all :) I reverted to the prev version.

I guess that the design of the columns in the `xor` example would be something like this:

| adv__1 | adv_2 | adv_3 | table_col_a | table_col_b | table_col_c |
| ------ | ----- | ----- | ----------- | ----------- | ----------- |
| cell_a     | cell_b     | cell_c     | 0           | 0           | 0           |
|        |       |       | 0           | 1           | 1           |
|        |       |       | 1           | 0           | 1           |
|        |       |       | 1           | 1           | 0           |

> That means cells a, b, and c must be on the same row of the table columns.

Why is that?

> Writing one gate that returns a vector of two elements adds an additional constraint that a and b have to be equal.

I still don't get why is that.  In the `xor` example, even though we created one big gate, we weren't constraining `cell_a`, `cell_b`, `cell_c` to be equal.

Reply:
When we do XOR, we are looking up for 
```
(a, b, c) ∈ {(0, 0, 0), (0, 1, 1), (1, 0, 1), (1, 1, 0)}
```
but not for
```
(a ∈ {0, 0, 1, 1}) ^(b ∈ {0, 1, 0, 1}) ^(c ∈ {0, 1, 1, 0})
```

`a=1, b=1, c=1` is an invalid XOR witness, but it is allowed in the second expression.

Reply:
I tried to create an example to reproduce the lookup to the same table column.

https://github.com/ChihChengLiang/halo2_examples/blob/bb94962a3a9a92394ff0b088fa431e8173f3ef57/src/main.rs#L128-L130

Reply:
Thanks cc. The `xor` example is fully clear to me. 

So trying to apply this concept to the `range_check` example.

| adv__1 | lookup_u8 |
| ------ | ----- | 
| u8_a   | 0     |
| u8_b   | 1      |   
|            |    2   |  
|            |     ...  |  
|            |   255  |  

1.

```
(u8_a, u8_b) ∈ {0, 1, 2, ..., 255}
```

This gate constrains `u8_a` and `ub_a` to be in the range **and** to be equal.

2.

```
(u8_a ∈ {0, 1, 2, ..., 255}) ^ (u8_b ∈ {0, 1, 2, ..., 255})
```

This gate constraints `u8_a` and `ub_a` to be in the range.

Is that correct?

Reply:
Yes, that's correct.